### ISSUE TITLE
`lighttpd` instrumentation crashes on `Adt`

### ISSUE THREAD

**Issue by kkysen** [immunant/c2rust#570]
`master` is failing to instrument [`lighttpd`](https://github.com/immunant/lighttpd-rust), panicking because `TyKind::Adt` is not handled in `ArgKind::from_type`/`ArgKind::from_type` is used in contexts it's not meant to handle.

```shell
> ./scripts/pdg.sh ../lighttpd/rust/
...
thread 'rustc' panicked at 'operand is not of integer-castable type: op = _4, ty = src::configfile::config_t, ty.kind(): Adt(src::configfile::config_t, [])', dynamic_instrumentation/src/arg.rs:30:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/core/src/panicking.rs:143:14
   2: c2rust_instrument::arg::ArgKind::from_type
             at /home/kkysen/work/rust/c2rust/dynamic_instrumentation/src/arg.rs:30:13
   3: c2rust_instrument::point::build::InstrumentationBuilder::arg_var
             at /home/kkysen/work/rust/c2rust/dynamic_instrumentation/src/point/build.rs:83:42
   4: c2rust_instrument::instrument::<impl rustc_middle::mir::visit::Visitor for c2rust_instrument::point::InstrumentationAdder>::visit_assign::{{closure}}
             at /home/kkysen/work/rust/c2rust/dynamic_instrumentation/src/instrument.rs:181:13
   5: c2rust_instrument::instrument::<impl rustc_middle::mir::visit::Visitor for c2rust_instrument::point::InstrumentationAdder>::visit_assign
             at /home/kkysen/work/rust/c2rust/dynamic_instrumentation/src/instrument.rs:190:17
   6: rustc_middle::mir::visit::Visitor::super_statement
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/compiler/rustc_middle/src/mir/visit.rs:381:25
   7: rustc_middle::mir::visit::Visitor::visit_statement
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/compiler/rustc_middle/src/mir/visit.rs:95:17
   8: rustc_middle::mir::visit::Visitor::super_basic_block_data
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/compiler/rustc_middle/src/mir/visit.rs:312:21
   9: rustc_middle::mir::visit::Visitor::visit_basic_block_data
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/compiler/rustc_middle/src/mir/visit.rs:84:17
  10: rustc_middle::mir::visit::Visitor::super_body
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/compiler/rustc_middle/src/mir/visit.rs:261:21
  11: rustc_middle::mir::visit::Visitor::visit_body
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/compiler/rustc_middle/src/mir/visit.rs:78:17
  12: c2rust_instrument::instrument::instrument_body
             at /home/kkysen/work/rust/c2rust/dynamic_instrumentation/src/instrument.rs:406:5
  13: c2rust_instrument::instrument::Instrumenter::instrument_fn
             at /home/kkysen/work/rust/c2rust/dynamic_instrumentation/src/instrument.rs:61:9
  14: c2rust_instrument::override_queries::{{closure}}
             at /home/kkysen/work/rust/c2rust/dynamic_instrumentation/src/main.rs:334:13
  15: core::ops::function::FnOnce::call_once
             at /rustc/1e12aef3fab243407f9d71ba9956cb2a1bf105d5/library/core/src/ops/function.rs:227:5
  16: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, &rustc_data_structures::steal::Steal<rustc_middle::mir::Body>>
  17: rustc_data_structures::stack::ensure_sufficient_stack::<(&rustc_data_structures::steal::Steal<rustc_middle::mir::Body>, rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, &rustc_data_structures::steal::Steal<rustc_middle::mir::Body>>::{closure#3}>
  18: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, &rustc_data_structures::steal::Steal<rustc_middle::mir::Body>>>
  19: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::mir_built
  20: rustc_mir_transform::check_unsafety::unsafety_check_result
  21: <rustc_mir_transform::check_unsafety::provide::{closure#0} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId)>>::call_once
  22: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId, &rustc_middle::mir::query::UnsafetyCheckResult>
  23: rustc_data_structures::stack::ensure_sufficient_stack::<(&rustc_middle::mir::query::UnsafetyCheckResult, rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, rustc_span::def_id::LocalDefId, &rustc_middle::mir::query::UnsafetyCheckResult>::{closure#3}>
  24: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, &rustc_middle::mir::query::UnsafetyCheckResult>>
  25: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::unsafety_check_result
  26: rustc_mir_transform::mir_const
  27: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, &rustc_data_structures::steal::Steal<rustc_middle::mir::Body>>
  28: rustc_data_structures::stack::ensure_sufficient_stack::<(&rustc_data_structures::steal::Steal<rustc_middle::mir::Body>, rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, &rustc_data_structures::steal::Steal<rustc_middle::mir::Body>>::{closure#3}>
  29: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, &rustc_data_structures::steal::Steal<rustc_middle::mir::Body>>>
  30: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::mir_const
  31: rustc_mir_transform::mir_promoted
  32: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, (&rustc_data_structures::steal::Steal<rustc_middle::mir::Body>, &rustc_data_structures::steal::Steal<rustc_index::vec::IndexVec<rustc_middle::mir::Promoted, rustc_middle::mir::Body>>)>
  33: rustc_data_structures::stack::ensure_sufficient_stack::<((&rustc_data_structures::steal::Steal<rustc_middle::mir::Body>, &rustc_data_structures::steal::Steal<rustc_index::vec::IndexVec<rustc_middle::mir::Promoted, rustc_middle::mir::Body>>), rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, (&rustc_data_structures::steal::Steal<rustc_middle::mir::Body>, &rustc_data_structures::steal::Steal<rustc_index::vec::IndexVec<rustc_middle::mir::Promoted, rustc_middle::mir::Body>>)>::{closure#3}>
  34: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_middle::ty::WithOptConstParam<rustc_span::def_id::LocalDefId>, (&rustc_data_structures::steal::Steal<rustc_middle::mir::Body>, &rustc_data_structures::steal::Steal<rustc_index::vec::IndexVec<rustc_middle::mir::Promoted, rustc_middle::mir::Body>>)>>
  35: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::mir_promoted
  36: <rustc_borrowck::provide::{closure#0} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId)>>::call_once
  37: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId, &rustc_middle::mir::query::BorrowCheckResult>
  38: rustc_data_structures::stack::ensure_sufficient_stack::<(&rustc_middle::mir::query::BorrowCheckResult, rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, rustc_span::def_id::LocalDefId, &rustc_middle::mir::query::BorrowCheckResult>::{closure#3}>
  39: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, &rustc_middle::mir::query::BorrowCheckResult>>
  40: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::mir_borrowck
  41: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#2}>
  42: rustc_interface::passes::analysis
  43: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), core::result::Result<(), rustc_errors::ErrorReported>>
  44: rustc_data_structures::stack::ensure_sufficient_stack::<(core::result::Result<(), rustc_errors::ErrorReported>, rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, (), core::result::Result<(), rustc_errors::ErrorReported>>::{closure#3}>
  45: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorReported>>>
  46: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
  47: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorReported>>
  48: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>
```

This is the code that panics: https://github.com/immunant/c2rust/blob/5fa0aec26c31983622a8cc833cfb43cac5d97908/dynamic_instrumentation/src/arg.rs#L20-L32

It seems to be from the [`TyKind::Adt`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/sty/enum.TyKind.html#variant.Adt) (an algebraic data-type, a `struct` in this case) not being handled.  We don't handle most `TyKind`s, though.  Why is that?  Or should such `TyKind`s not be passed to `ArgKind::from_type` in the first place?

I think there was a change in a previous PR that landed that @fw-immunant had mentioned might not handle all cases, but I'm not sure where that is.

**Comment by fw-immunant** [comment]
`instrument.rs` wants to "`// add instrumentation for load-from-address operations`" but I'm not sure exactly what we want to record in this situation. At present, for any assignments where the rvalue is a `Use` or `AddressOf`, and the rvalue place involves dereferencing any pointer, we pass the place to `add_load_instr`, which attempts to record the entire value at that place.

This fails when the value at the place is not compatible with our instrumentation functions, e.g. if it is a struct such as `src::configfile::config_t` as it is in the backtrace above.

What exactly do we want to record in this situation? The existing comment is confusing to me, because the code to instrument is in the context of assignment statements, but is commented as if it intends to instrument memory loads in general. Do we have to instrument these piecemeal (in assignments and every other context where they might appear), or is there a centralized place we can handle them (`visit_place`?), or is the comment misleading?

**Comment by aneksteind** [comment]
@fw-immunant we want to record the value of the addresses that are loaded from -- but we weren't checking to see that the type getting dereferenced was a pointer or reference in this case. I've put an insufficient fix here https://github.com/immunant/c2rust/pull/582 that may get us past this particular occurrence. We will need a more involved fix in general (https://github.com/immunant/c2rust/issues/578) to ensure the places getting dereferenced are what's actually passed to the hook. Right now, the dynamic analysis code incorrectly assumes it's always the base local.

The reason this code isn't in `visit_place` is because it lacks context to determine whether or not a load really happens (because not all dereferences in projections imply a load from an address, e.g. `_2 = &raw *_1`)

**Comment by kkysen** [comment]
Closed by #582.
